File: 1163. Last Substring in Lexicographical Order/Solution.java
class Solution {
    public String lastSubstring(String s) {
        int i = 0;
        for (int j = i + 1, k = 0; j + k < s.length(); ) {
            if (s.charAt(i + k) == s.charAt(j + k)) {
                k++;
            } else if (s.charAt(i + k) < s.charAt(j + k)) {
                i += k + 1;
                k = 0;
                if (i >= j) {
                    j = i + 1;
                }
            } else {
                j += k + 1;
                k = 0;
            }
        }
        return s.substring(i);
    }
}

File: 1163. Last Substring in Lexicographical Order/Solution.py
class Solution:
    def lastSubstring(self, s: str) -> str:
        i, j, k, n = 0, 1, 0, len(s)
        while j + k < n:
            if s[i + k] == s[j + k]:
                k += 1
            elif s[i + k] < s[j + k]:
                i += k + 1
                k = 0
                if i >= j:
                    j = i + 1
            else:
                j += k + 1
                k = 0
        return s[i:]
